cmd/agent/app/flags.go

```diff
@@ -22,69 +22,84 @@ package app
 )
 
-// Bind binds the agent builder to command line options
-func (b *Builder) Bind(flags *flag.FlagSet) {
+// CreateFlags returns a new FlagSet that defines parameters of the Builder where default
+// flag values are retrieved from the current values of the Builder's fields
+func (b *Builder) CreateFlags() *flag.FlagSet {
+	flagSet := &flag.FlagSet{}
+
 	for i := range b.Processors {
 		p := &b.Processors[i]
-		name := "processor." + string(p.Model) + "-" + string(p.Protocol) + "."
-		flags.IntVar(
-			&p.Workers,
-			name+"workers",
+		prefix := processorFlagPrefix(p)
+		flagSet.Int(
+			prefix+suffixWorkers,
 			p.Workers,
 			"how many workers the processor should run")
-		flags.IntVar(
-			&p.Server.QueueSize,
-			name+"server-queue-size",
+		flagSet.Int(
+			prefix+suffixServerQueueSize,
 			p.Server.QueueSize,
 			"length of the queue for the UDP server")
-		flags.IntVar(
-			&p.Server.MaxPacketSize,
-			name+"server-max-packet-size",
+		flagSet.Int(
+			prefix+suffixServerMaxPacketSize,
 			p.Server.MaxPacketSize,
 			"max packet size for the UDP server")
-		flags.StringVar(
-			&p.Server.HostPort,
-			name+"server-host-port",
+		flagSet.String(
+			prefix+suffixServerHostPort,
 			p.Server.HostPort,
 			"host:port for the UDP server")
 	}
-	flags.Var(
-		&stringSliceFlag{slice: &b.CollectorHostPorts},
-		"collector.host-port",
+	flagSet.String(
+		collectorHostPort,
+		strings.Join(b.CollectorHostPorts, ","),
 		"comma-separated string representing host:ports of a static list of collectors to connect to directly (e.g. when not using service discovery)")
-	flags.StringVar(
-		&b.HTTPServer.HostPort,
-		"http-server.host-port",
+	flagSet.String(
+		httpServerHostPort,
 		b.HTTPServer.HostPort,
-		"host:port of the http server (e.g. for /sampling and /baggage endpoint)")
-	flags.IntVar(
-		&b.DiscoveryMinPeers,
-		"discovery.min-peers",
+		"host:port of the http server (e.g. for /sampling point and /baggage endpoint)")
+	flagSet.Int(
+		discoveryMinPeers,
 		3,
 		"if using service discovery, the min number of connections to maintain to the backend")
-}
 
-type stringSliceFlag struct {
-	slice *[]string
+	return flagSet
 }
 
-// String formats the flag's value, part of the flag.Value interface.
-func (c *stringSliceFlag) String() string {
-	return fmt.Sprint(c.slice)
-}
+// InitFromViper builder properties
+func (b *Builder) InitFromViper(v *viper.Viper) {
+	// make sure that viper is initialized with default values for a case when CreateFlags is not
+	// passed to viper
+	b.CreateFlags().VisitAll(func(flag *flag.Flag) {
+		v.SetDefault(flag.Name, flag.DefValue)
+	})
 
-// Set sets the flag value, part of the flag.Value interface.
-func (c *stringSliceFlag) Set(value string) error {
-	if len(*(c.slice)) > 0 {
-		return fmt.Errorf("comma-separated flag already set: %v", *(c.slice))
+	for i := range b.Processors {
+		p := &b.Processors[i]
+		prefix := processorFlagPrefix(p)
+		p.Workers = v.GetInt(prefix + suffixWorkers)
```

Comment:
It does not work out of the box, I think struct field names have to match flags (or viper keys) to make it work.